File: apps/ios/CereBroTests/RefusalCodesTest.swift
import XCTest
@testable import CereBro

/// The refusals the server sends, told apart by CODE rather than by status.
///
/// The same contract web (`tests/app/refusalCodes.test.ts`) and Android
/// (`RefusalCodesTest.kt`) pin. Three different 429s now mean three different
/// things — the free-tier cap, a daily abuse ceiling, and slowapi's "slow
/// down" — and one of the wrong answers is manipulative: offering an upgrade
/// for a ceiling that is identical on every tier is selling a fix that is not
/// for sale.
///
/// The 403 tests are the ones that matter most here. iOS collapsed EVERY 403
/// into `.unauthorized` — "Your session expired. Please sign in again." — so
/// the verification gate would have told people to sign out of an account that
/// was working perfectly well.
final class RefusalCodesTest: XCTestCase {

    private func body(_ json: String) -> [String: Any]? {
        guard let data = json.data(using: .utf8) else { return nil }
        return try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }

    // MARK: - Parsed by code, never by status

    func testADailyCeilingIsRecognisedAndCarriesItsFeature() {
        let parsed = APIClient.dailyCeiling(from: body("""
        {"detail":{"code":"daily_ceiling","feature":"voice_tts",
        "message":"You've hit the daily limit for this.","limit":2000,
        "resets_at":"2026-08-24T00:00:00+00:00"}}
        """))
        XCTAssertEqual(parsed?.feature, "voice_tts")
        XCTAssertEqual(parsed?.limit, 2000)
        XCTAssertNotNil(parsed?.resetsAt)
    }

    func testTheFreeCapAndTheDailyCeilingDoNotParseAsEachOther() {
        // Same status, opposite remedy. A screen catching the wrong one would
        // offer to sell a fix that does not exist.
        let ceiling = """
        {"detail":{"code":"daily_ceiling","feature":"oracle_turn","limit":500}}
        """
        let cap = """
        {"detail":{"code":"free_daily_limit","limit":50,"used":50}}
        """
        XCTAssertNil(APIClient.freeLimit(from: body(ceiling)))
        XCTAssertNil(APIClient.dailyCeiling(from: body(cap)))
        XCTAssertNotNil(APIClient.dailyCeiling(from: body(ceiling)))
        XCTAssertNotNil(APIClient.freeLimit(from: body(cap)))
    }

    func testAnOrdinaryThrottleIsNeitherOfThem() {
        // slowapi's key is `error`, with no `detail`, and it means "slow down".
        let throttle = body("""
        {"error":"Rate limit exceeded: 10 per 1 minute"}
        """)
        XCTAssertNil(APIClient.freeLimit(from: throttle))
        XCTAssertNil(APIClient.dailyCeiling(from: throttle))
    }

    // MARK: - The 403 that is not a dead session

    func testTheVerificationGateIsRecognisedAndNamesItsFeature() {
        let parsed = APIClient.verificationRequired(from: body("""
        {"detail":{"code":"email_unverified","feature":"voice",
        "message":"Confirm your email address to use this."}}
        """))
        XCTAssertEqual(parsed?.feature, "voice")
        XCTAssertEqual(parsed?.message, "Confirm your email address to use this.")
    }

    func testAPlain403IsNotMistakenForAVerificationWall() {
        // A consent-gated 403 answers with a STRING detail and must not offer
        // to resend a verification email.
        let consent = body("""
        {"detail":"AI memory is switched off in your privacy settings."}
        """)
        XCTAssertNil(APIClient.verificationRequired(from: consent))
    }

    func testTheVerificationMessageIsNotTheSessionExpiredOne() {
        // The whole point of the case existing. `.unauthorized` says "Your
        // session expired. Please sign in again." — advice that would have sent
        // somebody to sign out of a perfectly good account.
        let info = VerificationRequiredInfo(
            message: "Confirm your email address to use this.", feature: "plans")
        let error = APIError.verificationRequired(info)
        XCTAssertEqual(error.errorDescription, info.message)
        XCTAssertNotEqual(error.errorDescription, APIError.unauthorized.errorDescription)
    }

    // MARK: - Copy that has to be right in every timezone

    func testTheResetIsRenderedLocallyRatherThanCalledMidnight() {
        // The window is UTC, so "midnight" is wrong for most of the world — in
        // India these clear at 05:30 local.
        let info = DailyCeilingInfo(
            message: "m", feature: "voice_tts", limit: 2000,
            resetsAt: Date(timeIntervalSince1970: 1_787_500_800))
        XCTAssertFalse(info.resetText.contains("midnight"))
        XCTAssertTrue(info.resetText.hasPrefix("resets at "))
    }

    func testWithNoInstantItNamesUTCRatherThanGuessing() {
        let info = DailyCeilingInfo(
            message: "m", feature: "voice_tts", limit: 2000, resetsAt: nil)
        XCTAssertEqual(info.resetText, "resets at midnight UTC")
    }
}

File: apps/android/app/src/test/java/com/cerebrozen/app/net/RefusalCodesTest.kt
package com.cerebrozen.app.net

import kotlinx.coroutines.test.runTest
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The refusals the server sends, told apart by CODE rather than by status.
 *
 * Three different 429s now mean three different things — the free-tier cap, a
 * daily abuse ceiling, and slowapi's "slow down" — and one of the wrong answers
 * is manipulative: offering an upgrade for a ceiling that is identical on every
 * tier is selling a fix that is not for sale.
 *
 * The queue tests are the ones that were load-bearing. None of these exceptions
 * is an [Session.ApiException], so before [Session.RefusalException] existed
 * they fell through [Outbox]'s catch-all — the branch meaning "no connectivity
 * at all" — and the write was queued. A refusal would then be retried on every
 * drain, forever, while the person was told it was saved and would sync. That
 * was already true of the free-tier cap before today.
 */
class RefusalCodesTest {

    private class FakeStore(vararg seed: Pair<String, String>) : Session.Store {
        val map = mutableMapOf(*seed)
        override fun getString(key: String) = map[key]
        override fun putString(key: String, value: String) { map[key] = value }
        override fun remove(key: String) { map.remove(key) }
        override fun keys() = map.keys.toSet()
    }

    private val tokens =
        """{"access_token":"a1","refresh_token":"r1","token_type":"bearer"}"""

    private fun respondingWith(status: Int, body: String) {
        Session.resetForTest(FakeStore("refresh_token" to "r1")) { url, _, _, _, _, _ ->
            if (url.endsWith("/auth/refresh")) 200 to tokens else status to body
        }
    }

    // ── Parsed by code, never by status ───────────────────────────────────

    @Test
    fun aDailyCeilingArrivesTypedAndCarriesItsFeature() = runTest {
        respondingWith(
            429,
            """{"detail":{"code":"daily_ceiling","feature":"voice_tts",""" +
                """"message":"You've hit the daily limit for this.","limit":2000,""" +
                """"resets_at":"2026-08-24T00:00:00Z"}}""",
        )
        try {
            Session.api("/voice/tts", "POST", JSONObject().put("text", "hi"))
            throw AssertionError("expected DailyCeilingException")
        } catch (e: Session.DailyCeilingException) {
            assertEquals("voice_tts", e.feature)
            assertEquals(2000, e.limit)
            assertEquals("2026-08-24T00:00:00Z", e.resetsAtUtc)
        }
    }

    @Test
    fun aDailyCeilingIsNotTheFreeCap() = runTest {
        // Same status, opposite remedy: the cap means "upgrade and this goes
        // away", the ceiling means "come back tomorrow". A screen that caught
        // the wrong one would offer to sell a fix that does not exist.
        respondingWith(
            429,
            """{"detail":{"code":"daily_ceiling","feature":"oracle_turn","limit":500}}""",
        )
        try {
            Session.api("/oracle/messages", "POST", JSONObject().put("text", "hi"))
            throw AssertionError("expected DailyCeilingException")
        } catch (e: Session.DailyCeilingException) {
            assertFalse(e is Session.FreeLimitException)
        }
    }

    @Test
    fun anUnconfirmedAddressArrivesTypedAndNamesTheFeature() = runTest {
        respondingWith(
            403,
            """{"detail":{"code":"email_unverified","feature":"voice",""" +
                """"message":"Confirm your email address to use this."}}""",
        )
        try {
            Session.api("/voice/tts", "POST", JSONObject().put("text", "hi"))
            throw AssertionError("expected VerificationRequiredException")
        } catch (e: Session.VerificationRequiredException) {
            // Named so a screen can say WHICH thing is waiting rather than
            // showing a generic wall.
            assertEquals("voice", e.feature)
            assertTrue(e.message!!.contains("Confirm your email"))
        }
    }

    @Test
    fun anOrdinaryThrottleStaysAPlainApiException() = runTest {
        // slowapi's key is `error`, with no `detail`, and it means "slow down".
        respondingWith(429, """{"error":"Rate limit exceeded: 10 per 1 minute"}""")
        try {
            Session.api("/auth/signup", "POST", JSONObject())
            throw AssertionError("expected ApiException")
        } catch (e: Session.ApiException) {
            assertEquals(429, e.code)
            assertTrue(e.message!!.contains("Rate limit exceeded"))
        }
    }

    @Test
    fun aPlain403IsNotMistakenForAVerificationWall() = runTest {
        // Only the code makes it one. A consent-gated 403 answers differently
        // and must not offer to resend a verification email.
        respondingWith(403, """{"detail":"AI memory is switched off in your privacy settings."}""")
        try {
            Session.api("/users/me/memory/x", "PATCH", JSONObject())
            throw AssertionError("expected ApiException")
        } catch (e: Session.ApiException) {
            assertEquals(403, e.code)
        }
    }

    // ── The queue must not keep a decision ────────────────────────────────

    @Test
    fun everyRefusalSharesTheBaseTheQueueChecks() {
        // Outbox branches on `Session.RefusalException`, so a future refusal
        // type that forgets to extend it silently becomes retryable-forever
        // again. This is the assertion that makes the base load-bearing rather
        // than decorative.
        val free = Session.FreeLimitException("m", 50, 50, "2026-08-24T00:00:00Z")
        val ceiling = Session.DailyCeilingException("m", "voice_tts", 2000, "2026-08-24T00:00:00Z")
        val verify = Session.VerificationRequiredException("m", "voice")

        assertTrue(free is Session.RefusalException)
        assertTrue(ceiling is Session.RefusalException)
        assertTrue(verify is Session.RefusalException)
        // And a transport failure is NOT one — that is the case the queue exists
        // for, and treating it as a refusal would drop somebody's write.
        assertFalse(Session.ApiException(503, "down") is Session.RefusalException)
    }

    @Test
    fun aRefusedWriteIsNotQueued() = runTest {
        // `send` used to reach its catch-all — "no connectivity at all" — and
        // enqueue the item, so the person was told it was saved and would sync
        // while the server had already decided otherwise.
        val store = FakeStore("refresh_token" to "r1")
        Session.resetForTest(store) { url, _, _, _, _, _ ->
            if (url.endsWith("/auth/refresh")) 200 to tokens
            else 403 to """{"detail":{"code":"email_unverified","feature":"plans"}}"""
        }
        Outbox.clear()

        try {
            Outbox.send("/moods", JSONObject().put("mood", "Good"))
            throw AssertionError("expected the refusal to reach the caller")
        } catch (e: Session.VerificationRequiredException) {
            // The caller sees it, which is the point.
        }
        assertEquals(0, Outbox.pending().size)
    }

    @Test
    fun aGenuineOutageIsStillQueued() = runTest {
        // The other half: the queue must still do its job. Without this the
        // test above could be satisfied by never queueing anything.
        val store = FakeStore("refresh_token" to "r1")
        Session.resetForTest(store) { url, _, _, _, _, _ ->
            if (url.endsWith("/auth/refresh")) 200 to tokens
            else throw java.net.UnknownHostException("no dns")
        }
        Outbox.clear()

        Outbox.send("/moods", JSONObject().put("mood", "Good"))
        assertEquals(1, Outbox.pending().size)
    }
}
